### PR TITLE
Introduce dynamic network fee

### PR DESCRIPTION
**Pull request by Kubuxu**

### CODE REVIEW

File: chain/store/basefee.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+	"github.com/filecoin-project/lotus/build"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/specs-actors/actors/abi"
+	"github.com/filecoin-project/specs-actors/actors/abi/big"
+	"golang.org/x/xerrors"
+)
+
+func (cs *ChainStore) ComputeBaseFee(ctx context.Context, ts *types.TipSet) (abi.TokenAmount, error) {
+	zero := abi.NewTokenAmount(0)
```
Comment (magik6k) [inline chain/store/basefee.go:14]: There's `abi/big.Zero()`

Comment (Kubuxu) [reply chain/store/basefee.go:14]: thre is `big.Zero` but I don't import it

File: chain/types/message.go [hunk 2]
```diff
@@ -138,7 +139,7 @@ func (m *Message) ValidForBlockInclusion(minGas int64) error {
 		return xerrors.New("'Value' field cannot be greater than total filecoin supply")
 	}
 
-	if m.GasPrice.LessThan(big.Zero()) {
+	if m.GasFeeCap.LessThan(big.Zero()) {
```
Comment (magik6k) [inline chain/types/message.go:142]: GasPremium Needs to be checked too

File: chain/vm/burn.go [hunk 3]
```diff
@@ -37,8 +50,48 @@ func ComputeGasOutputs(gasUsed, gasLimit int64) (int64, int64) {
...
+		MinerTip:           big.Zero(),
+		Refund:             big.Zero(),
+	}
+
+	baseFeeToPay := baseFee
+	if baseFee.Cmp(feeCap.Int) > 0 {
+		baseFeeToPay = feeCap
+		out.MinerPenalty = big.Mul(big.Sub(baseFee, feeCap), gasUsedBig)
+	}
+	out.BaseFeeBurn = big.Mul(baseFeeToPay, big.NewInt(gasUsed))
```
Comment (magik6k) [inline chain/vm/burn.go:74]:
```suggestion
	out.BaseFeeBurn = big.Mul(baseFeeToPay, gasUsedBig)
```

File: chain/messagepool/selection.go [hunk 4]
```diff
@@ -30,39 +30,55 @@ type msgChain struct {
...
-		next := mp.createMessageChains(actor, mset, ts)
+		next := mp.createMessageChains(actor, mset, baseFee, ts)
 		chains = append(chains, next...)
 	}
 
 	// 2. Sort the chains
 	sort.Slice(chains, func(i, j int) bool {
 		return chains[i].Before(chains[j])
 	})
+	if len(chains) != 0 && chains[0].gasPerf < 0 {
```
Comment (magik6k) [inline chain/messagepool/selection.go:75]: Looks like we may still end up including chains with negative gasPerf here

File: chain/store/basefee.go (hunk 1 above)
Comment (magik6k) [reply chain/store/basefee.go:28]: L9?

File: chain/messagepool/selection.go [hunk 5]
```diff
@@ -88,7 +88,7 @@ func (mp *MessagePool) selectMessages(curTs, ts *types.TipSet) ([]*types.SignedM
 	last := len(chains)
 	for i, chain := range chains {
 		// does it fit in the block?
-		if chain.gasLimit <= gasLimit {
```
Comment (magik6k) [inline chain/messagepool/selection.go:75]: Any reason to not check for >=

File: chain/messagepool/selection.go [hunk 6]
```diff
@@ -72,7 +88,7 @@ func (mp *MessagePool) selectMessages(curTs, ts *types.TipSet) ([]*types.SignedM
...
+		if chain.gasLimit <= gasLimit && chain.gasPerf > 0 {
```
Comment (magik6k) [inline chain/messagepool/selection.go:91]: Why not check for >=?

**Pull request merged**